ci: Only write docker build images to Cirrus cache

maflcko opened the pull request:
The `DOCKER_BUILD_CACHE_ARG` env var holds the options on how to use cache providers. Storing the image layers is useful for the Cirrus cache provider, because it offers 10GB per runner (https://cirrus-runners.app/setup/#speeding-up-the-cache). The cached image layers can help to avoid issues when the upstream package manager infra (apt native, apt llvm, pip, apk, git clone, ...) has outages or network issues.

However, on the GitHub Actions cache provider, a *total* cache of 10GB is offered for the whole repo. This cache must be shared with the depends cache, and the ccache, as well as the previous releases cache. So it is already full and trying to put the docker build layers into it will lead to an overflow.

Fix it by only writing to the docker cache on Cirrus.

Also, `DOCKER_BUILD_CACHE_ARG` requires a `shellcheck disable=SC2086` on the full build command. Fix that as well by using `shlex.split` from Python on just this variable.

Commit:
ci: Move buildx command to python script

This has a few benefits:

* The shellcheck SC2086 warning is disabled for the whole command, but
  is only needed for the DOCKER_BUILD_CACHE_ARG env var.  So in Python,
  only pass this one env var to shlex.split() for proper word splitting.
* Future logic improvements can be implemented in Python.

The comments are moved, which can be checked via the git options:
--color-moved=dimmed-zebra --color-moved-ws=ignore-all-space

Commit:
ci: Only write docker build images to Cirrus cache

Other cache providers offer too little space for this to be useful.

cedwies commented on ci/test/02_run_container.sh:
```diff
@@ -10,30 +10,6 @@ export CI_IMAGE_LABEL="bitcoin-ci-test"
-      --tag="${CONTAINER_NAME}" \
-      $DOCKER_BUILD_CACHE_ARG \
-      "${BASE_READ_ONLY_DIR}"
-
```
Maybe a short comment is helpful for future maintainers?
`# The Docker build step has moved into 02_run_container.py (see PR #33639)`

maflcko replied:
I think it is fine without comment, because the files are numbered in order and not too many, so one can just read/grep them all to search what they are looking for.

cedwies reviewed (commented):
Approach ACK

maflcko commented on ci/test/02_run_container.py:
```diff
@@ -17,6 +18,18 @@ def run(cmd, **kwargs):
+        P = int(run(['nproc'], stdout=subprocess.PIPE).stdout)
+        M = min(P, int(os.getenv("MAKEJOBS").lstrip("-j")))
+        cpus = sorted(random.sample(range(P), M))
+        return [f"--cpuset-cpus={','.join(map(str, cpus))}"]
```
I don't think `MAKEJOBS=-j0` was ever supported by the CI. It will eventually be passed down to the test runner, which fails:

```
test/functional/test_runner.py", line 714, in __init__
    assert num_tests_parallel >= 1
           ^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

So I think I'll leave this as-is for now, to keep the code move-only, as the goal is not to change it. It can be changed in a separate PR, if there is need.

Just as a hint: If you want to use `nproc`, with the current CI system, you can just omit all settings, because it is picked by default:

https://github.com/bitcoin/bitcoin/blob/e744fd1249bf9577274614eaf3997bf4bbb612ff/ci/test/00_setup_env.sh#L37-L38

cedwies reviewed (commented):
utACK fa29ecd
Reviewed the code and the new CI flow. The change to limit cache writes to Cirrus (while still reading cache on GHA) and moving the build step into Python with `shlex.split(os.getenv("DOCKER_BUILD_CACHE_ARG", ""))` looks good to me. I don’t see any unintended behavior changes beyond what is described in the PR.

l0rinc commented on .github/actions/configure-docker/action.yml:
```diff
@@ -48,8 +48,9 @@ runs:
-        if [[ ${{ github.event_name }} == "push" && ${{ github.ref_name }} == ${{ github.event.repository.default_branch }} ]]; then
+        # Only when using the Cirrus cache provider and this is a push to the default
+        # branch, add --cache-to to save the cache.
+        if [[ ${{ inputs.cache-provider }} == 'cirrus' && ${{ github.event_name }} == "push" && ${{ github.ref_name }} == ${{ github.event.repository.default_branch }} ]]; then
```
slightly tangential, but maybe there's room for optimizing the docker layers by putting the variable parts as the last layers, using ARGs over ENVs, combining multiple layers, pruning old ones, etc.

l0rinc commented on ci/test/02_run_container.py:
```diff
@@ -45,6 +58,30 @@ def main():
+    if not os.getenv("DANGER_RUN_CI_ON_HOST"):
+        CI_IMAGE_LABEL = "bitcoin-ci-test"
+
+        maybe_cache_arg = shlex.split(os.getenv("DOCKER_BUILD_CACHE_ARG", ""))
```
Is this related to the mentioned https://www.shellcheck.net/wiki/SC2086? Because we need split to handle quotes and escaping properly?

l0rinc reviewed (approved):
Concept ACK, I understand that you already have an ACK but please consider my suggestions

Commit:
ci: Remove unused MAYBE_CPUSET

The option is currently unused. If it is used again in the future, it
could trivially be added back.

Also, the logic is just a single undocumented python command one-liner.

So remove it for now.

maflcko commented on ci/test/02_run_container.py:
```diff
@@ -45,6 +58,30 @@ def main():
+        # Using buildx is required to properly load the correct driver, for use with registry caching. Neither build, nor BUILDKIT=1 currently do this properly
+        cmd_build = ["docker", "buildx", "build"]
+        cmd_build += [
+            f"--file={os.getenv('BASE_READ_ONLY_DIR')}/ci/test_imagefile",
```
Ah nice, changed all

l0rinc commented on ci/test/02_run_container.py:
```diff
@@ -45,6 +58,30 @@ def main():
+        cmd_build += [os.getenv("BASE_READ_ONLY_DIR")]
+
+        print(f"Building {os.getenv('CONTAINER_NAME')} image tag to run in")
+        run(cmd_build)
```
In the followup please consider avoiding concats (e.g. via `*list`), as shown above

l0rinc reviewed (approved):
Code review ACK fabe0e07de1ad2f26da62f3ebe0e9be3f939b1f8

Let's see how much slowdown this would cause without caching.

maflcko commented:
> Let's see how much slowdown this would cause without caching.

I don't think this has any measurable effect on the runtime. The only benefit could be avoiding intermittent network errors.

m3dwards commented:
ACK fabe0e07de1ad2f26da62f3ebe0e9be3f939b1f8

This will save a lot of space and help to prevent the other caches from getting evicted which are much more important than the docker build layer caches. Should only slow down the jobs by a few minutes at most and hopefully be a net positive as the other more important caches (like ccache) will not have been evicted.

Ran on master on my own fork.

willcl-ark reviewed (approved):
ACK fabe0e07de1ad2f26da62f3ebe0e9be3f939b1f8

Agree with the rationale here and the changes look correct to me. There is a run on-going on a fork (non Cirrus runners) at https://github.com/m3dwards/bitcoin/actions/runs/18709746522

...and the [caches in that repo](https://github.com/m3dwards/bitcoin/actions/caches) do not contain docker build images as intended.

maflcko commented:
Hmm, so I did a fresh run in a repo with a clean GHA cache, and it now says:

>  Approaching total cache storage limit (8.89 GB of 10 GB Used) 

So I guess we are still close to overlapping the cache and certainly will reach cache churn with more than one push to the default branch.
